db/db_flush_test.cc

```diff
@@ -740,7 +740,88 @@ class TestFlushListener : public EventListener {
   DBFlushTest* test_;
 };
 
-// RocksDB lite does not support GetLiveFiles()
+TEST_F(
+    DBFlushTest,
+    FixUnrecoverableWriteDuringAtomicFlushWaitUntilFlushWouldNotStallWrites) {
+  Options options = CurrentOptions();
+  options.atomic_flush = true;
+  options.disable_auto_compactions = true;
+  // To simulate a real-life crash where we can't flush during db's shutdown
+  options.avoid_flush_during_shutdown = true;
+  CreateAndReopenWithCF({"cf1"}, options);
+
+  // Write some initial data so we have something to atomic-flush later
+  // triggered by `GetLiveFiles()`
+  WriteOptions write_opts;
+  write_opts.disableWAL = true;
+  ASSERT_OK(Put(1, "k1", "v1", write_opts));
+
+  // To mock a write stall condition where atomic flush will release the lock
+  // to wait, during which we can write to the db
+  bool mocked_write_stall_condition = false;
```

Comment:
I took a look and realized I chose such a bad name for `mocked_write_stall_condition`. It's intended to be the one returned from [GetWriteStallConditionAndCause](https://github.com/facebook/rocksdb/blob/main/db/column_family.cc#L873), which don't seem to relate to write_controller. Regardless, this is a too much of a detail to expose to reviewer for he/she to learn about the UT from scratch. 

I removed the added sync point and set some options.

I also realized this UT will greatly benefit from a visualization of the problem so it could be understood easily why I did something in test. I attached that here too, though not nearly as good as whiteboard (sigh)

Let me know if see other opportunities to improve the UT and make it easier to understand. 

![UT and problem description](https://user-images.githubusercontent.com/83968999/222300078-8aafdd06-4b5f-4528-ab1d-e4c55a91bf33.png)

Reply:
> It's intended to be the one returned from [GetWriteStallConditionAndCause](https://github.com/facebook/rocksdb/blob/main/db/column_family.cc#L873), which don't seem to relate to write_controller

Understood. This was non-obvious to me because typically `GetWriteStallConditionAndCause()` is used for deciding whether to stall writes, which is `WriteController`'s responsibility. However `WaitUntilFlushWouldNotStallWrites` uses it to predict whether write stall would occur under possible future conditions. (Embarrassingly that was added by me yet I still don't know it #3381)